pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGateFunctors.hpp

```diff
@@ -27,13 +27,108 @@ namespace {
 using namespace Pennylane::Util;
 using Kokkos::kokkos_swap;
 using Pennylane::Gates::GateOperation;
+using Pennylane::LightningKokkos::Util::ControlBitPatterns;
+using Pennylane::LightningKokkos::Util::generateBitPatterns;
+using Pennylane::LightningKokkos::Util::parity_2_offset;
+using Pennylane::LightningKokkos::Util::reverseWires;
 using Pennylane::LightningKokkos::Util::vector2view;
 } // namespace
 /// @endcond
 
 namespace Pennylane::LightningKokkos::Functors {
+template <class PrecisionT, class FuncT> class applyNCNFunctor {
+    using KokkosComplexVector = Kokkos::View<Kokkos::complex<PrecisionT> *>;
+    using KokkosIntVector = Kokkos::View<std::size_t *>;
+    using MemberType = Kokkos::TeamPolicy<>::member_type;
 
-template <class PrecisionT, class FuncT> class applyNC1Functor {
+    Kokkos::View<Kokkos::complex<PrecisionT> *> arr;
+    const FuncT core_function;
+    KokkosIntVector indices;
+    KokkosIntVector parity;
+    KokkosIntVector rev_wires;
+    KokkosIntVector rev_wire_shifts;
+    std::size_t dim;
+
+  public:
+    template <class ExecutionSpace>
+    applyNCNFunctor([[maybe_unused]] ExecutionSpace exec,
+                    Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
+                    std::size_t num_qubits,
+                    const std::vector<std::size_t> &controlled_wires,
+                    const std::vector<bool> &controlled_values,
+                    const std::vector<std::size_t> &wires, FuncT core_function_)
+        : arr(arr_), core_function(core_function_) {
+        const std::size_t two2N =
+            exp2(num_qubits - wires.size() - controlled_wires.size());
+        dim = exp2(wires.size());
+        const auto &[parity_, rev_wires_] =
+            reverseWires(num_qubits, wires, controlled_wires);
+        parity = parity_;
+        std::vector<std::size_t> indices_ =
+            generateBitPatterns(wires, num_qubits);
```

Comment:
Is it possible to parallelize the `generateBitPatterns` method within the LK frame work? If yes, we can add a `TODO` for it?

Reply:
Yes it should be possible, I've added a comment in the util function.